File: java/src/GH/GH220521/프로그래머스/타겟넘버.java
package GH.GH220521.프로그래머스;

public class 타겟넘버 {
    static int result;
    public static void main (String[] arg){
        int[] n = {1, 1, 1, 1, 1};
        System.out.print(solution(n,3));
    }

    private static int solution(int[] numbers, int target){
        dfs(numbers,target,0,0);
        return result;
    }

    private static void dfs (int[] n, int target, int idx, int sum){
        if (idx >= n.length){
            if (sum == target) result++;
            return;
        }

        int add = sum + n[idx];
        dfs (n, target, idx+1, add);

        int sub = sum - n[idx];
        dfs (n, target, idx+1, sub);
    }
}

File: SeoungHoon(JS)/Programmers/타겟넘버.js
function recursive(target, array, index) {
    var result = 0;
    for (let i = index; i < array.length ; i++) {
        var temp = target;
        temp -= array[i]
        if (temp == 0){
            result++;
        } else if (temp < 0){
            continue;
        } else {
            result += recursive(temp, array, i+1)     }
    }
    return result;
}

function solution(numbers, target) {
    var answer = 0;
    var k = (numbers.reduce((pre, cur) => pre + cur) - target) / 2;
    
    return recursive(k, numbers, 0);
}
